src/core/qgsfields.h

```diff
@@ -347,6 +347,15 @@ class CORE_EXPORT  QgsFields
     % End
 #endif
 
+    /**
+     * Returns TRUE if this field is a read-only field or should be treated as
+     * such given its origin. This is the case for providers which support
+     * generated fields for instance, or when the origin of the field is
+     * OriginJoin or OriginExpression
+     * \since QGIS 3.38
+     */
+    bool isFieldReadOnly( int fieldIdx ) const;
```

Comment:
Can you add similar SIP logic as fieldOrigin so that a KeyError is raised for invalid indexes?